### PR TITLE
Add page number pagination to ListMyEvents and EventSearch

### PR DESCRIPTION
**Pull request by nabramow**
This PR adds pageNumber pagination to `ListMyEvents` and `EventSearch` while maintaining the existing `pageToken` pagination.

Discussed with @bakeiro and decided to do it this way if the backend part won't take too long.

**Backend checklist**
- [ x ] Formatted my code by running `ruff check --select I --fix . && ruff check . && ruff format .` in `app/backend`
- [ x ] Added tests for any new code or added a regression test if fixing a bug
- [ x ] All tests pass
- [ x ] Run the backend locally and it works
- [ N/A ] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

Closes #4939

### CODE REVIEW

**Comment by nabramow** [comment]
Also halp why is my deployment failing, I'm not able to see it since we don't have an account for me there.

File: app/backend/src/couchers/servicers/events.py [hunk 1]
```diff
@@ -1043,13 +1048,19 @@ def ListMyEvents(self, request, context, session):
...
+        occurrences = (
+            occurrences.offset(offset).limit(page_size) if request.page_number else occurrences.limit(page_size + 1)
+        )
         occurrences = session.execute(occurrences).scalars().all()
 
         return events_pb2.ListMyEventsRes(
             events=[event_to_pb(session, occurrence, context) for occurrence in occurrences[:page_size]],
             next_page_token=str(millis_from_dt(occurrences[-1].end_time)) if len(occurrences) > page_size else None,
+            page_number=page_number,
+            total_items=total_items,
```
Comment (bakeiro) [inline app/backend/src/couchers/servicers/events.py:1062]: I believe (I can be wrong here, just an possible idea)

changing the return of this endpoint, could lead to a different proto object, leading to a different type in the frontend, and maybe that could be the issue of the build failing?

**Comment by bakeiro** [comment]
> Also halp why is my deployment failing, I'm not able to see it since we don't have an account for me there.

the result of the build it's in Gitlab... do you need an account to see it? anyway, by executing `yarn build` in the directory `app/web` is causing some issues... maybe that's the root of the bug

**Comment by nabramow** [comment]
> > Also halp why is my deployment failing, I'm not able to see it since we don't have an account for me there.
> 
> the result of the build it's in Gitlab... do you need an account to see it? anyway, by executing `yarn build` in the directory `app/web` is causing some issues... maybe that's the root of the bug

No, when I click through it says I need an account and Aapeli said it costs to add seats. `yarn build` was a good idea though, that gave me a clue, thanks!

File: app/backend/src/couchers/servicers/events.py (hunk 1 above)
Comment (nabramow) [reply app/backend/src/couchers/servicers/events.py:1062]: The issue ended up being that the `MyEvents` tests was throwing an error as it was missing the `pageNumber` and `totalItems` in the api response. I handled this by making these keys optional in the response in the protos. Then we can just use them if we want to, but don't have to.

File: app/backend/src/couchers/servicers/events.py [hunk 2]
```diff
@@ -1043,13 +1048,19 @@ def ListMyEvents(self, request, context, session):
...
+        occurrences = (
+            occurrences.offset(offset).limit(page_size) if request.page_number else occurrences.limit(page_size + 1)
+        )
         occurrences = session.execute(occurrences).scalars().all()
 
         return events_pb2.ListMyEventsRes(
             events=[event_to_pb(session, occurrence, context) for occurrence in occurrences[:page_size]],
             next_page_token=str(millis_from_dt(occurrences[-1].end_time)) if len(occurrences) > page_size else None,
+            page_number=page_number if request.page_number else None,
+            total_items=total_items if request.page_number else None,
```
Comment (aapeliv) [inline app/backend/src/couchers/servicers/events.py]: Please return this regardless of if `page_number` is set, given that it's computed anyway.

File: app/backend/src/couchers/servicers/search.py [hunk 3]
```diff
@@ -660,10 +665,14 @@ def EventSearch(self, request, context, session):
...
+        total_items = session.execute(select(func.count()).select_from(statement.subquery())).scalar()
+        # Apply pagination by page number
+        statement = statement.offset(offset).limit(page_size) if request.page_number else statement.limit(page_size + 1)
         occurrences = session.execute(statement).scalars().all()
 
         return search_pb2.EventSearchRes(
             events=[event_to_pb(session, occurrence, context) for occurrence in occurrences[:page_size]],
             next_page_token=(str(millis_from_dt(occurrences[-1].end_time)) if len(occurrences) > page_size else None),
+            page_number=page_number if request.page_number else None,
+            total_items=total_items if request.page_number else None,
```
Comment (aapeliv) [inline app/backend/src/couchers/servicers/search.py]: ditto

File: app/proto/events.proto [hunk 4]
```diff
@@ -280,12 +280,15 @@ message ListMyEventsReq {
 
   uint32 page_size = 5;
   string page_token = 6;
+  optional uint32 page_number = 9;
```
Comment (aapeliv) [inline app/proto/events.proto]: Everything is `optional`, no need for it in proto3.

Please change to

```proto3
oneof pagination {
  string page_token = 6;
  uint32 page_number = 9;
}
```

File: app/proto/events.proto [hunk 5]
```diff
@@ -280,12 +280,15 @@ message ListMyEventsReq {
...
 }
 
 message ListMyEventsRes {
   repeated Event events = 1;
 
   string next_page_token = 2;
+  optional uint32 page_number = 3;
```
Comment (aapeliv) [inline app/proto/events.proto]: Do you need to return the page number if it's just what was in the input?

File: app/proto/search.proto [hunk 6]
```diff
@@ -198,10 +198,13 @@ message EventSearchReq {
 
   uint32 page_size = 10;
   string page_token = 11;
+  optional uint32 page_number = 18;
```
Comment (aapeliv) [inline app/proto/search.proto]: Again `oneof` thing

File: app/proto/search.proto [hunk 7]
```diff
@@ -198,10 +198,13 @@ message EventSearchReq {
...
 }
 
 message EventSearchRes {
   repeated org.couchers.api.events.Event events = 1;
 
   string next_page_token = 2;
+  optional uint32 page_number = 3;
+  optional uint32 total_items = 4;
```
Comment (aapeliv) [inline app/proto/search.proto]: Please remove all `optional`s.

So proto2 syntax had this required/optional thing. proto3 has everything be optional. It's a bit silly, optional/required was a good thing, but they simplified the language. In proto3 if something's missing it gets a default value (e.g. 0, empty string, etc, depending on type). proto2 was more powerful, you could define the default if it was missing.

**Review by aapeliv — CHANGES_REQUESTED** [review]
Looking good! I put in a couple small change requests.

In general I think pagination can cause some performance issues, but there shouldn't be that many events so it should be fine. But if we wanted to do this with e.g. users, we might need to think about it some more!

File: app/proto/events.proto (hunk 5 above)
Comment (nabramow) [reply app/proto/events.proto]: Good point, probably not. Fixed.

File: app/proto/search.proto (hunk 7 above)
Comment (nabramow) [reply app/proto/search.proto]: Thanks for the background! I've made the changes now and updated the tests.

File: app/proto/events.proto [hunk 8]
```diff
@@ -280,12 +280,14 @@ message ListMyEventsReq {
 
   uint32 page_size = 5;
   string page_token = 6;
+  uint32 page_number = 9;
```
Comment (aapeliv) [inline app/proto/events.proto]: Please change to

```
oneof pagination {
  string page_token = 6;
  uint32 page_number = 9;
}
```

aka just wrap both inside the `oneof pagination`

File: app/proto/search.proto [hunk 9]
```diff
@@ -198,10 +198,12 @@ message EventSearchReq {
 
   uint32 page_size = 10;
   string page_token = 11;
+  uint32 page_number = 18;
```
Comment (aapeliv) [inline app/proto/search.proto]: Same here

**Review by aapeliv — CHANGES_REQUESTED** [review]
just one small thing!

File: app/proto/events.proto (hunk 8 above)
Comment (nabramow) [reply app/proto/events.proto]: Okay just pushed this! Had to adjust the events.py and seach.py a bit for the oneof so the tests would still pass :-)

File: app/backend/src/couchers/servicers/events.py [hunk 10]
```diff
@@ -978,9 +978,23 @@ def SetEventAttendance(self, request, context, session):
...
+        # Check which field in the 'pagination' oneof is set
+        pagination_type = request.WhichOneof("pagination")
         page_size = min(MAX_PAGINATION_LENGTH, request.page_size or MAX_PAGINATION_LENGTH)
-        # the page token is a unix timestamp of where we left off
-        page_token = dt_from_millis(int(request.page_token)) if request.page_token else now()
+
+        if pagination_type == "page_token":
+            # the page token is a unix timestamp of where we left off
+            page_token = dt_from_millis(int(request.page_token))
+            page_number = 1
```
Comment (aapeliv) [inline app/backend/src/couchers/servicers/events.py]: I don't think these `page_number`s are used now outside the `pagination_type == "page_number"` case

File: app/backend/src/couchers/servicers/search.py [hunk 11]
```diff
@@ -660,10 +673,13 @@ def EventSearch(self, request, context, session):
                 EventOccurrence.start_time.desc()
             )
 
-        statement = statement.limit(page_size + 1)
+        total_items = session.execute(select(func.count()).select_from(statement.subquery())).scalar()
+        # Apply pagination by page number
+        statement = statement.offset(offset).limit(page_size) if request.page_number else statement.limit(page_size + 1)
```
Comment (aapeliv) [inline app/backend/src/couchers/servicers/search.py:670]: It seems `offset` is not defined if you have `pagination_type == "page_token"`

File: app/backend/src/couchers/servicers/events.py (hunk 10 above)
Comment (nabramow) [reply app/backend/src/couchers/servicers/events.py]: You are right, apologies. I realize I ran the tests before regenerating the protos and assumed the error was this.

File: app/backend/src/couchers/servicers/search.py (hunk 11 above)
Comment (nabramow) [reply app/backend/src/couchers/servicers/search.py:670]: I changed it back, turns out I did not need these changes.

**Review by aapeliv — APPROVED** [review]
LGTM if the tests pass!

**Pull request merged**